### ISSUE TITLE
CCSD 'LinAlgError: Singular matrix' for custom Hamiltonian

### ISSUE THREAD

**Issue by sweetpotat5** [pyscf/pyscf#1981]
Hi, I was trying to implement the ccsd calculation on the two-dimensional Hubbard model on a square lattice following Figure 5 on page 33 in this paper: https://arxiv.org/pdf/1701.08223.pdf.

(1) when running the code, the ccsd output reports the error 'LinAlgError: Singular matrix'. The ccsd correlation energy from the beginning is very small and the optimization behavior is abnormal. 

(2) when testing with the code from the paper, I'm getting exactly the same energy result from both 'cc.CCSD(mf).run()' and 
'cc.UCCSD(mf).run()' code, which shouldn't be the case. Why do ccsd and uccsd generate the same result?

Can anyone explain why I'm getting such behaviors? Any help would be appreciated!

**the test code I use for the 2d hubbard case:**
# 2d hubbard square
import numpy as np
from pyscf import gto, scf, ao2mo, cc

mol = gto.Mole(verbose=4)
n = 4
mol.nelectron = n*n
mol.spin = 0
t, u = 1, 8

mf = scf.RHF(mol)

h1 = np.zeros((n*n,n*n))
for i in range(n): # apply pbc
    for j in range(n):
        h1[i*n + j, ((i+1) % n)*n + j] = h1[((i+1) % n)*n + j, i*n + j] = t
        h1[i*n + j, i*n + (j+1) % n] = h1[i*n + (j+1) % n, i*n + j] = t

mf.get_hcore = lambda *args: h1
mf.get_ovlp = lambda *args: np.eye(n*n)
mf._eri = np.zeros((n*n, n*n, n*n, n*n)) # two-electron integral
for i in range(n*n):
    mf._eri[i,i,i,i] = u

mf.run()
cc.CCSD(mf).run()

**the complete output:**
[CONFIG] conf_file None
[INPUT] verbose = 4
[INPUT] num. atoms = 0
[INPUT] num. electrons = 16
[INPUT] charge = 0
[INPUT] spin (= nelec alpha-beta = 2S) = 0
[INPUT] symmetry False subgroup None
[INPUT] Mole.unit = angstrom
[INPUT] Symbol           X                Y                Z      unit          X                Y                Z       unit  Magmom

nuclear repulsion = 0
number of shells = 0
number of NR pGTOs = 0
number of NR cGTOs = 0
basis = sto-3g
ecp = {}
CPU time:        24.26

******** <class 'pyscf.scf.hf.RHF'> ********
method = RHF
initial guess = minao
damping factor = 0
level_shift factor = 0
DIIS = <class 'pyscf.scf.diis.CDIIS'>
diis_start_cycle = 1
diis_space = 8
SCF conv_tol = 1e-09
SCF conv_tol_grad = None
SCF max_cycles = 50
direct_scf = True
direct_scf_tol = 1e-13
chkfile to save SCF result = /var/folders/r4/km4mckcn0x97wsb979r6ws8r0000gn/T/tmppq35m4zs
max_memory 4000 MB (current use 0 MB)
Set gradient conv threshold to 3.16228e-05
No atom found in mol. Use 1e initial guess
Initial guess from hcore.

WARN: HOMO -2.68058944013731e-16 == LUMO -1.24335583315498e-16

init E= 10.0525093356927
  HOMO = 3.59792387867847  LUMO = 4.30865784508607
cycle= 1 E= 17.3067486914261  delta_E= 7.25  |g|= 10.3  |ddm|= 4.75
  HOMO = 3.1672015674873  LUMO = 4.90789847287264
cycle= 2 E= 29.2178327720607  delta_E= 11.9  |g|= 15.2  |ddm|= 5.68
  HOMO = 3.97420998600601  LUMO = 4.00077256117936
cycle= 3 E= 11.5631920992419  delta_E= -17.7  |g|= 5.63  |ddm|=  3.5
  HOMO = 3.66248109277178  LUMO = 4.30900691732218
cycle= 4 E= 13.5656850837006  delta_E=    2  |g|= 7.53  |ddm|= 4.71
  HOMO = 4.01797752905939  LUMO = 4.03717393287724
cycle= 5 E= 11.0486463136645  delta_E= -2.52  |g|= 4.95  |ddm|= 2.89
  HOMO = 3.95618442750099  LUMO = 3.97497351114688
cycle= 6 E= 11.0432418141595  delta_E= -0.0054  |g|= 4.86  |ddm|= 3.26
  HOMO = 3.96584956239016  LUMO = 4.0587511876615
cycle= 7 E= 11.5189701352313  delta_E= 0.476  |g|= 5.22  |ddm|= 3.97
  HOMO = 4.0101425823  LUMO = 4.01468930601442
cycle= 8 E= 10.7407968735476  delta_E= -0.778  |g|= 4.55  |ddm|= 3.14
  HOMO = 4.00832148278664  LUMO = 4.01103235704372
cycle= 9 E= 11.1455846807804  delta_E= 0.405  |g|=  4.7  |ddm|= 2.88
  HOMO = 3.99744589230339  LUMO = 4.00177669809386
cycle= 10 E= 10.8099596027907  delta_E= -0.336  |g|= 4.75  |ddm|= 2.87
  HOMO = 4.01298042509293  LUMO = 4.02925465035913
cycle= 11 E= 11.2517379735604  delta_E= 0.442  |g|= 4.89  |ddm|= 2.88
  HOMO = 3.98231281956881  LUMO = 4.01329133772819
cycle= 12 E= 11.2098944628251  delta_E= -0.0418  |g|= 4.88  |ddm|= 2.44
  HOMO = 3.99489382152202  LUMO = 3.9982943423453
cycle= 13 E= 11.0244351294833  delta_E= -0.185  |g|= 4.87  |ddm|= 4.44
  HOMO = 3.99430477955457  LUMO = 4.05810874944195
cycle= 14 E= 11.5520643021007  delta_E= 0.528  |g|= 5.37  |ddm|= 3.19
  HOMO = 4.00796204268747  LUMO = 4.05377599251264
cycle= 15 E= 11.5056811637485  delta_E= -0.0464  |g|= 5.23  |ddm|= 3.38
  HOMO = 3.95814595259361  LUMO = 4.05216191920819
cycle= 16 E= 11.7508740075687  delta_E= 0.245  |g|= 5.55  |ddm|= 2.99
  HOMO = 3.97200002440826  LUMO = 4.01080900829007
cycle= 17 E= 11.0609291889124  delta_E= -0.69  |g|= 4.85  |ddm|= 3.82
  HOMO = 3.9948459736793  LUMO = 4.01396585045287
cycle= 18 E= 11.2305670023867  delta_E= 0.17  |g|= 4.75  |ddm|= 3.55
  HOMO = 3.99416776924603  LUMO = 4.00378294970055
cycle= 19 E= 11.0270534013359  delta_E= -0.204  |g|= 4.72  |ddm|= 3.81
  HOMO = 3.9845777230276  LUMO = 4.02461971205565
cycle= 20 E= 11.1701887556787  delta_E= 0.143  |g|= 4.92  |ddm|= 4.49
  HOMO = 3.96718202926566  LUMO = 4.06768284965324
cycle= 21 E= 11.6064008092407  delta_E= 0.436  |g|= 5.38  |ddm|= 3.95

WARN: HOMO 3.99698745309865 == LUMO 3.99785745685594

cycle= 22 E= 10.496642438321  delta_E= -1.11  |g|= 4.49  |ddm|= 3.06
  HOMO = 4.01264020068208  LUMO = 4.0152799477125
cycle= 23 E= 10.960103989809  delta_E= 0.463  |g|= 4.65  |ddm|= 3.04
  HOMO = 4.00815822569556  LUMO = 4.01000026942284
cycle= 24 E= 10.946939440779  delta_E= -0.0132  |g|= 4.66  |ddm|= 2.85
  HOMO = 3.93735194657335  LUMO = 4.0897732489204
cycle= 25 E= 11.6777343675514  delta_E= 0.731  |g|= 5.43  |ddm|=  3.2
  HOMO = 3.96327413480211  LUMO = 4.03459014033308
cycle= 26 E= 11.4595825680628  delta_E= -0.218  |g|= 5.13  |ddm|= 4.02
  HOMO = 3.98953984254256  LUMO = 4.01799445067726
cycle= 27 E= 11.4132036889623  delta_E= -0.0464  |g|= 5.11  |ddm|= 3.43
  HOMO = 3.97937797954841  LUMO = 4.0078483408333
cycle= 28 E= 11.3655939250582  delta_E= -0.0476  |g|= 5.16  |ddm|=  3.9
  HOMO = 3.98466703120481  LUMO = 4.02903010384483
cycle= 29 E= 11.2603244090115  delta_E= -0.105  |g|= 4.85  |ddm|= 2.96
  HOMO = 3.97333707019017  LUMO = 4.0330306906042
cycle= 30 E= 11.4695271676078  delta_E= 0.209  |g|= 5.15  |ddm|= 3.83
  HOMO = 3.99680052279806  LUMO = 4.03522294199108
cycle= 31 E= 11.3074676083365  delta_E= -0.162  |g|= 4.94  |ddm|= 3.56
  HOMO = 3.94021713922485  LUMO = 4.07951338062357
cycle= 32 E= 11.5518450326562  delta_E= 0.244  |g|= 5.56  |ddm|= 3.79
  HOMO = 3.96590778926306  LUMO = 4.02117370872563
cycle= 33 E= 11.3131644761982  delta_E= -0.239  |g|= 5.02  |ddm|=  3.2
  HOMO = 3.95781978054112  LUMO = 4.06324800278288
cycle= 34 E= 11.6747813116501  delta_E= 0.362  |g|= 5.39  |ddm|= 3.68
  HOMO = 3.91523590083477  LUMO = 4.09233924062645
cycle= 35 E= 11.7530246807105  delta_E= 0.0782  |g|= 5.45  |ddm|= 4.06
  HOMO = 3.90066476402649  LUMO = 4.0982891716527
cycle= 36 E= 11.9598310721751  delta_E= 0.207  |g|= 5.81  |ddm|= 3.17
  HOMO = 3.98781371405256  LUMO = 4.00403623543199
cycle= 37 E= 12.6723070812132  delta_E= 0.712  |g|= 6.95  |ddm|= 2.83
  HOMO = 3.93149851891148  LUMO = 4.00250035196073
cycle= 38 E= 11.5856697439759  delta_E= -1.09  |g|= 5.24  |ddm|=    4
  HOMO = 3.95098415467878  LUMO = 4.03138722057128
cycle= 39 E= 11.6685846478553  delta_E= 0.0829  |g|= 5.33  |ddm|= 3.92
  HOMO = 3.98259171707581  LUMO = 4.03355811884787
cycle= 40 E= 11.2976312587817  delta_E= -0.371  |g|=  5.1  |ddm|= 3.37
  HOMO = 3.94114877713188  LUMO = 4.00743082256635
cycle= 41 E= 11.5147699839905  delta_E= 0.217  |g|= 5.13  |ddm|= 3.89
  HOMO = 4.00745410898668  LUMO = 4.04467578921962
cycle= 42 E= 11.3520993195025  delta_E= -0.163  |g|= 5.05  |ddm|= 2.69
  HOMO = 3.993818326448  LUMO = 4.00402401158583
cycle= 43 E= 11.2754424742278  delta_E= -0.0767  |g|= 4.92  |ddm|= 4.02
  HOMO = 3.96615321953209  LUMO = 4.11619333121021
cycle= 44 E= 11.8042703171578  delta_E= 0.529  |g|= 5.58  |ddm|= 3.98
  HOMO = 3.99636032697075  LUMO = 4.00136847062098
cycle= 45 E= 11.1717548674325  delta_E= -0.633  |g|= 4.72  |ddm|=  3.5
  HOMO = 3.93838663162499  LUMO = 4.08431846515519
cycle= 46 E= 11.6438388044678  delta_E= 0.472  |g|= 5.54  |ddm|= 3.95
  HOMO = 3.97359590669664  LUMO = 3.99735509177812
cycle= 47 E= 11.2333497079371  delta_E= -0.41  |g|= 4.95  |ddm|= 3.88
  HOMO = 3.92826997832941  LUMO = 4.07473059733919
cycle= 48 E= 11.7891899418901  delta_E= 0.556  |g|= 5.55  |ddm|= 3.89
  HOMO = 3.98532627586893  LUMO = 4.0194299227794
cycle= 49 E= 11.1485005813588  delta_E= -0.641  |g|= 4.81  |ddm|= 3.93
  HOMO = 3.9813628544349  LUMO = 4.00880611948739
cycle= 50 E= 11.2707117962685  delta_E= 0.122  |g|= 5.09  |ddm|= 3.59
SCF not converged.
SCF energy = 11.2707117962685

******** <class 'pyscf.cc.ccsd.CCSD'> ********
CC2 = 0
CCSD nocc = 8, nmo = 16
max_cycle = 50
direct = 0
conv_tol = 1e-07
conv_tol_normt = 1e-05
diis_space = 6
diis_start_cycle = 0
diis_start_energy_diff = 1e+09
max_memory 4000 MB (current use 0 MB)
Init t2, MP2 energy = 32.7728652115917  E_corr(MP2) 21.5021534153232
Init E_corr(CCSD) = 18.8629553951056
cycle = 1  E_corr(CCSD) = -368397.932794775  dE = -368416.796  norm(t1,t2) = 1.05244e+06
cycle = 2  E_corr(CCSD) = -1.8179958792228e-43  dE = 368397.933  norm(t1,t2) = 1.24338e+14
cycle = 3  E_corr(CCSD) = 0  dE = 1.81799588e-43  norm(t1,t2) = 112.23
cycle = 4  E_corr(CCSD) = 0  dE = 0  norm(t1,t2) = 112.23
Warning: <pyscf.gto.mole.Mole object at 0x297080810> must be initialized before calling SCF.
Initialize <pyscf.gto.mole.Mole object at 0x297080810> in <pyscf.scf.hf.RHF object at 0x297011e10>
cycle = 5  E_corr(CCSD) = 0  dE = 0  norm(t1,t2) = 112.23
cycle = 6  E_corr(CCSD) = 0  dE = 0  norm(t1,t2) = 112.23
cycle = 7  E_corr(CCSD) = 0  dE = 0  norm(t1,t2) = 112.23
cycle = 8  E_corr(CCSD) = 18.8629584780099  dE = 18.8629585  norm(t1,t2) = 112.23
cycle = 9  E_corr(CCSD) = -20.3737686739861  dE = -39.2367272  norm(t1,t2) = 1.05244e+06
cycle = 10  E_corr(CCSD) = -19.5086920145341  dE = 0.865076659  norm(t1,t2) = 3397.57
cycle = 11  E_corr(CCSD) = -17.8056719287611  dE = 1.70302009  norm(t1,t2) = 1389.85

WARN:  diis singular, eigh(h) [-1.21546421e-01  2.10612433e-06  1.68261676e+01  8.00677721e+04
  2.35754062e+05  8.81733720e+06  1.10763845e+12]

---------------------------------------------------------------------------
LinAlgError                               Traceback (most recent call last)
Cell In[27], line 25
     22     mf._eri[i,i,i,i] = u
     24 mf.run()
---> 25 cc.CCSD(mf).run()

File ~/anaconda3/lib/python3.11/site-packages/pyscf/lib/misc.py:599, in StreamObject.run(self, *args, **kwargs)
    592 '''
    593 Call the kernel function of current object.  `args` will be passed
    594 to kernel function.  `kwargs` will be used to update the attributes of
    595 current object.  The return value of method run is the object itself.
    596 This allows a series of functions/methods to be executed in pipe.
    597 '''
    598 self.set(**kwargs)
--> 599 self.kernel(*args)
    600 return self

File ~/anaconda3/lib/python3.11/site-packages/pyscf/cc/ccsd.py:1068, in CCSDBase.kernel(self, t1, t2, eris)
   1067 def kernel(self, t1=None, t2=None, eris=None):
-> 1068     return self.ccsd(t1, t2, eris)

File ~/anaconda3/lib/python3.11/site-packages/pyscf/cc/ccsd.py:1083, in CCSDBase.ccsd(self, t1, t2, eris)
   1079 if eris is None:
   1080     eris = self.ao2mo(self.mo_coeff)
   1082 self.converged, self.e_corr, self.t1, self.t2 = \
-> 1083         kernel(self, eris, t1, t2, max_cycle=self.max_cycle,
   1084                tol=self.conv_tol, tolnormt=self.conv_tol_normt,
   1085                verbose=self.verbose, callback=self.callback)
   1086 self._finalize()
   1087 return self.e_corr, self.t1, self.t2

File ~/anaconda3/lib/python3.11/site-packages/pyscf/cc/ccsd.py:84, in kernel(mycc, eris, t1, t2, max_cycle, tol, tolnormt, verbose, callback)
     82 t1, t2 = t1new, t2new
     83 t1new = t2new = None
---> 84 t1, t2 = mycc.run_diis(t1, t2, istep, normt, eccsd-eold, adiis)
     85 eold, eccsd = eccsd, mycc.energy(t1, t2, eris)
     86 log.info('cycle = %d  E_corr(%s) = %.15g  dE = %.9g  norm(t1,t2) = %.6g',
     87          istep+1, name, eccsd, eccsd - eold, normt)

File ~/anaconda3/lib/python3.11/site-packages/pyscf/cc/ccsd.py:1195, in CCSDBase.run_diis(self, t1, t2, istep, normt, de, adiis)
   1191 if (adiis and
   1192     istep >= self.diis_start_cycle and
   1193     abs(de) < self.diis_start_energy_diff):
   1194     vec = self.amplitudes_to_vector(t1, t2)
-> 1195     t1, t2 = self.vector_to_amplitudes(adiis.update(vec))
   1196     logger.debug1(self, 'DIIS for step %d', istep)
   1197 return t1, t2

File ~/anaconda3/lib/python3.11/site-packages/pyscf/lib/diis.py:237, in DIIS.update(self, x, xerr)
    235 else:
    236     self._xprev = None # release memory first
--> 237     self._xprev = xnew = self.extrapolate(nd)
    239     self._store('xprev', xnew)
    240     if 'xprev' not in self._buffer:  # not incore

File ~/anaconda3/lib/python3.11/site-packages/pyscf/lib/diis.py:264, in DIIS.extrapolate(self, nd)
    262     except numpy.linalg.linalg.LinAlgError as e:
    263         logger.warn(self, ' diis singular, eigh(h) %s', w)
--> 264         raise e
    265 logger.debug1(self, 'diis-c %s', c)
    267 xnew = None

File ~/anaconda3/lib/python3.11/site-packages/pyscf/lib/diis.py:261, in DIIS.extrapolate(self, nd)
    259 else:
    260     try:
--> 261         c = numpy.linalg.solve(h, g)
    262     except numpy.linalg.linalg.LinAlgError as e:
    263         logger.warn(self, ' diis singular, eigh(h) %s', w)

File ~/anaconda3/lib/python3.11/site-packages/numpy/linalg/linalg.py:409, in solve(a, b)
    407 signature = 'DD->D' if isComplexType(t) else 'dd->d'
    408 extobj = get_linalg_error_extobj(_raise_linalgerror_singular)
--> 409 r = gufunc(a, b, signature=signature, extobj=extobj)
    411 return wrap(r.astype(result_t, copy=False))

File ~/anaconda3/lib/python3.11/site-packages/numpy/linalg/linalg.py:112, in _raise_linalgerror_singular(err, flag)
    111 def _raise_linalgerror_singular(err, flag):
--> 112     raise LinAlgError("Singular matrix")

LinAlgError: Singular matrix

**Comment by verena-neufeld** [comment]
Hi @sweetpotat5 , thanks for your questions. Have you seen this example? 
https://github.com/pyscf/pyscf/blob/211979741467b2f5bc3ea685e12e9f5c097af451/examples/cc/40-ccsd_custom_hamiltonian.py
From a quick glance it seems that you might need a line that reformats the eri correctly like https://github.com/pyscf/pyscf/blob/211979741467b2f5bc3ea685e12e9f5c097af451/examples/cc/40-ccsd_custom_hamiltonian.py#L31 .
For UCCSD vs RCCSD, I'm not sure. Have you tested running UHF first and then UCCSD?

**Comment by tberkel** [comment]
You might need to make sure that your UHF initial guess breaks spin
symmetry so that you get a solution that is different from RHF.

**Comment by sweetpotat5** [comment]
Hi @verena-neufeld and @tberkel, 
(1) For the 2d Hubbard case: I've tried adding the suggested line 'mf._eri = ao2mo.restore(8, eri, n)' into the code, and I've also tested different symmetries (1,4, and 8). Unfortunately, the issue still exists with or without the code.
(2) For the CCSD vs UCCSD case: I tried the broken-symmetry methods from https://github.com/pyscf/pyscf.github.io/blob/master/examples/scf/32-break_spin_symm.py
and it worked. The example mentioned that 'The degeneracy is destroyed by zeroing out the beta 1s,2s components'. I wonder if this is the only way to break the symmetry (i.e. we can only zero out beta 1s 2s for all models) ?

Since broken-symmetry works for the 1d case, I also tried to use it on the 2d case. In the 4x4 Hubbard model, I can get quite an accurate energy for u=2 (0.4% relative error). As u increases, the accuracy quickly decreases from u=4 (8.5%) to u=6 (12.4%). Also, my UCCSD calculation cannot converge for u>=4 cases. Though I'm not getting the singular matrix error anymore, for the u=8 case I observe an oscillation behavior for the correlation energy as the graph shows. Maybe I need to play around with the parameters?

![pic1](https://github.com/pyscf/pyscf/assets/75410395/10f4b0fe-6103-4fc1-8f77-4734e4a4e4b3)

Thank you for the quick response and the solutions!
